### PR TITLE
Fix addDeveloperProduct & checkDeveloperProductName. addGamepass feature

### PR DESCRIPTION
**Pull request by emsar69**
Fixed addDeveloperProduct & checkDeveloperProductName. Added 'addGamepass' function.

### CODE REVIEW

File: lib/games/checkDeveloperProductName.js [hunk 1]
```diff
@@ -23,17 +23,25 @@ exports.func = (args) => {
...
     if (res.statusCode !== 200) {
       throw new Error('You are not logged in')
     } else {
-      return JSON.parse(res.body)
+      const data = JSON.parse(res.body)
+      let exists = false
+      data.forEach(product => {
+        if(product.name == args.productName || product.id == productId){
+          exists = true
+          return
```
Comment (Neztore) [inline lib/games/checkDeveloperProductName.js:41]: This returns undefined and the typing says it returns a boolean.
Would a better approach be:
```js
for (const product of data) {
   if (...) {
       return true

}
}
return false
```?

File: lib/games/checkDeveloperProductName.js [hunk 2]
```diff
@@ -23,17 +23,25 @@ exports.func = (args) => {
   const productId = parseInt(args.productId) ? parseInt(args.productId) : 0
 
   return http({
-    url: '//www.roblox.com/places/check-developerproduct-name?universeId=' + universeId + '&developerProductId=' + productId + '&developerProductName=' + args.productName + '&_=1515792139751',
+    url: `//apis.roblox.com/developer-products/v1/universes/${universeId}/developerproducts?pageNumber=1&pageSize=2147483647`,
```
Comment (Neztore) [inline lib/games/checkDeveloperProductName.js:26]: What is the meaning of the page size here? Why is it a large arbitrary number? how does this work when the number of pages is greater than 1? Could this use a pagination wrapper

File: lib/games/addGamepass.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * @param {number} priceInRobux - The price of the product.
+ * @param {string=} description - The description of the gamepass.
+ * @returns {Promise<GamepassAddResult>}
+ * @example const noblox = require("noblox.js")
+ * // Login using your cookie
+ * noblox.addGamepass(1, "A Gamepass", 100, "A cool item.")
+**/
+
+const nextFunction = (jar, token, universeId, name, priceInRobux, description) => {
+  const FormData = require('form-data');
```
Comment (Neztore) [inline lib/games/addGamepass.js:23]: Don't require modules in functions like this.
This does not need a new dependency - I believe the new fetch API has a built in way of building these, and they are built manually elsewhere in the library.

File: lib/games/addGamepass.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      jar: jar,
+      body: goofy,
+      headers: {
+        ...goofy.getHeaders(),
+        'X-CSRF-TOKEN': token
+      },
+      resolveWithFullResponse: true
+    }
+  }).then((res) => {
+    console.log(res)
```
Comment (Neztore) [inline lib/games/addGamepass.js:43]: console.log shouldn't be there

File: lib/games/addDeveloperProduct.js [hunk 5]
```diff
@@ -59,7 +59,7 @@ const nextFunction = (jar, token, universeId, name, priceInRobux, description) =
 exports.func = (args) => {
   const jar = args.jar
 
-  return getGeneralToken({ jar }).then((xcsrf) => {
+  return getGeneralToken({ jar: jar }).then((xcsrf) => {
```
Comment (Neztore) [inline lib/games/addDeveloperProduct.js:62]: Not sure what the point of this change is

**Review by Neztore — CHANGES_REQUESTED** [review]
Thank you for your PR, I've highlighted a few issues.
Your changes to checkDeveloperProductName would be breaking changes but we intend to move to a new major version - so that isn't an issue. I agree with your API changes there; returning a boolean is the best API.

I am not comfortable adding a dependency for form data as I believe it is done manually elsewhere and there are built in APIs for doing so (it is url encoded I believe so URLSearchParams could work)

**Comment by Neztore** [comment]
Adding help wanted if anyone wants to help out and resolve the above

File: lib/games/checkDeveloperProductName.js (hunk 2 above)
Comment (Regalijan) [reply lib/games/checkDeveloperProductName.js:26]: Per my testing, Roblox does not use cursors for this endpoint. The number specified for the page size is (2^31)-1 (the largest possible signed 32bit int), which happens to be the largest page size Roblox supports. Now, whether the api will actually return that many results, I don't know.

I agree that a pagination wrapper would be preferable to this though, given this is already undocumented to start with and I don't really expect this to work in practice (but maybe somehow it does).

**Comment by Noxirity** [comment]
Just seen this pr after I posted a fix for the same function https://github.com/noblox/noblox.js/pull/806

**Comment by Neztore** [comment]
Closing this as it is stale. Please open a new PR if above are fixed

**Pull request closed without merging**